### PR TITLE
Add typing definitions and add usage in modules - mainly for documentaion purposes

### PR DESCRIPTION
**Pull request by DBCerigo**

### CODE REVIEW

File: kotsu/typing.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+# The results from a validation run on a particular model.
+# Each Results dict will become one row of a combined results dataframe. Each key within the
+# Results dict will become a column of that dataframe, with values for matching keys across Results
+# put into the same matching column.
+# An example result from a 3-fold cross validation might be:
+# `{"average_score": _, "standard_deviation_scores": _, "fold_1_score": _, "fold_2_score"...}`
+Results = Dict[str, Union[float, str]]
+
+# Model's within kostu are not constrained in any way. It is up to the user to write Model's that
```
Comment (andrewsutjahjo) [inline kotsu/typing.py]: nit: plural of Model is Models

**Review by andrewsutjahjo — APPROVED** [review]
Nice one, only a nitpick on the documentation side

File: kotsu/typing.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+# have an interface compatible with their Validations.
+Model = Any
+
+# A Validation is a callable (e.g. function) that takes a Model, and returns a Results.
+Validation = Callable[[Any], Results]
```
Comment (ali-tny) [inline kotsu/typing.py:21]: 🐼  `Validation = Callable[[Model], Results]`

**Review by ali-tny — CHANGES_REQUESTED** [review]
looks great - just requesting changes cause I think we should also type the Registries with these new classes (which would make the inline typehints on line 36/37 of `kotsu/run.py` unnecessary), whis is a perfect use case for [Generic types](https://mypy.readthedocs.io/en/stable/generics.html)

**Review by benjamincerigo — APPROVED** [review]
Yeah this look much clearer. 

Nice one.

File: kotsu/typing.py (hunk 1 above)
Comment (DBCerigo) [reply kotsu/typing.py]: 40011cb

File: kotsu/typing.py (hunk 2 above)
Comment (DBCerigo) [reply kotsu/typing.py]: d88c564

**Comment by DBCerigo** [comment]
@ali-tny I had a stab at the `Generic` - seems like I did it right, but I tried adding in some incorrect type hinting (like swapping `Validation` for `Model` in some function sigs in `run` module) but `mypy` didn't throw anything :/ Guess there's a lot of `Any` going about so. In any case, if you could quick review it, and check you think I did it right at least, that would be appreciated thanks!

**Review by ali-tny — APPROVED** [review]
looks great! nice one

re efficacy - yeah the Model stuff is never gonna catch much cause Model is Any, so we don't care what the type is - but it is doing work on the Validation, given that you don't have to manually type the validation on line 36 of run. if you set change it to `Validation = str` in the typing file, you'll see a mypy error for line 61 of run (the only place where the validation is actually called!), showing that the type of Validation (as a callable) is correctly tracked through. Likewise if the signature of a validation was untyped or typed differently to required, it'll complain

**Comment by DBCerigo** [comment]
> looks great! nice one
> 
> re efficacy - yeah the Model stuff is never gonna catch much cause Model is Any, so we don't care what the type is - but it is doing work on the Validation, given that you don't have to manually type the validation on line 36 of run. if you set change it to `Validation = str` in the typing file, you'll see a mypy error for line 61 of run (the only place where the validation is actually called!), showing that the type of Validation (as a callable) is correctly tracked through. Likewise if the signature of a validation was untyped or typed differently to required, it'll complain

Nice nice. Lol I guess doing `validation: Model = ...` when `Model = Any` isn't going to throw much... derp. 

Thanks for the learnings and the reviews

**Pull request merged**